pandas/core/window_/aggregators.py

```diff
@@ -0,0 +1,136 @@
+import abc
+from typing import Optional
+
+import numpy as np
+
+from pandas._typing import Scalar
+
+
+class BaseAggregator(abc.ABC):
+    """Interface to return the current value of the rolling aggregation at the current step"""
+    def __init__(self, values: np.ndarray) -> None:
+        self.values = values
+
+    @abc.abstractmethod
+    def query(
+        self, start: int, stop: int, previous_start: int, previous_end: int
+    ) -> Scalar:
+        """
+        Computes the result of an aggregation for values that are between
+        the start and stop indices
+
+        Parameters
+        ----------
+        start : int
+            Current starting index
+
+        stop : int
+            Current stopping index
+
+        previous_start : int
+            Prior starting index
+
+        previous_end : int
+            Prior ending index
+
+        Returns
+        -------
+        Scalar
+            A scalar value that is the result of the aggregation.
+
+        """
+
+
+class AggKernel(abc.ABC):
+    """Interface that computes the aggregation value"""
+    def __init__(self):
+        self.count = 0
+
+    @abc.abstractmethod
+    def finalize(self):
+        """Return the final value of the aggregation."""
+
+
+class UnaryAggKernel(AggKernel):
+    @abc.abstractmethod
+    def step(self, value) -> None:
+        """Update the state of the aggregation with `value`."""
+
+
+class SubtractableAggregator(BaseAggregator):
+    def __init__(self, values: np.ndarray, agg: AggKernel) -> None:
+        super().__init__(values)
+        self.agg = agg
+
+    def query(
+        self, start: int, stop: int, previous_start: int, previous_end: int
+    ) -> Scalar:
+        """Compute a value based on changes in bounds."""
+        if previous_start == -1 and previous_end == -1:
+            for value in self.values[start:stop]:
+                self.agg.step(value)
+        else:
+            for value in self.values[previous_start:start]:
+                self.agg.invert(value)
+            for value in self.values[previous_end:stop]:
+                self.agg.step(value)
+        return self.agg.finalize()
+
+
+class Sum(UnaryAggKernel):
+    def __init__(self) -> None:
+        super().__init__()
+        self.total = 0
+
+    def step(self, value) -> None:
+        if value is not None:
+            self.count += 1
+            self.total += value
+
+    def invert(self, value) -> None:
+        """Used only in subtractable kernels."""
+        if value is not None:
+            self.count -= 1
+            self.total -= value
+
+    def finalize(self) -> Optional[int]:
```

Comment:
why is this optional?

Reply:
Because you can potentially aggregate an empty array.